programs/itools/doc/idate.1
.\" -*- nroff -*-
.\" $Id: idate.1 11011 2009-03-19 09:20:05Z thamer $
.\"
.\" NOTE: Please adjust .TH date whenever modifying the manpage.
.\"
.TH IDATE "1" "January 05, 2005" "idate" "User Commands"
.\"
.SH NAME
idate \- A Gregorian/Meladi to/from Hijri/Islamic date converter
.\"
.SH SYNOPSIS
.B idate
.RB [ \--gregorian
.RI   yyyymmdd ]
.RB [ \--hijri
.RI   yyyymmdd ]
.RB [ \--simple]
.RB [ \--umm_alqura ]
.RB [ \--help ]
.SH DESCRIPTION
The \fBidate\fR program is a Gregorian to Hijri (and vice-versa)
date converter.  The application uses and offers multiple calculation
methods with not all of them agreeing at all times.  The reason for
this multiplicity is due to not having one agreed upon method and
so various entities develop and advocate their calculations.
.PP
\fBidate\fR is able to comprehend and calculate both pre-epoch or
pre-Hijrah, denoted as "B.H", as well as post-epoch or post-Hijrah,
denoted as "A.H", dates.  \fBidate\fR also utilizes Gregorian's
pre-epoch "B.C" and post-epoch "A.D" dates and notes them per its
output.  When entering pre-epoch years, negative numbers ought to be
utilized.
.PP
\fBidate\fR when run without any command-line options uses the host
machine's current Gregorian date and converts it to Hijri.
.SH OPTIONS
\fBidate\fR follows the usual GNU command line syntax, with long options
starting with two dashes (`-').  A summary of all options is noted below:
.TP
.B \-h, \-\-help
Show summary of options
.TP
.B \-g, \-\-gregorian \fIyyyymmdd\fR
Specify the Gregorian date to be converted where 'y' stands for
year, 'm' for month and 'd' for day
.TP
.B \-hi, \-\-hijri \fIyyyymmdd\fR
Specify the Hijri date to be converted where 'y' stands for
year, 'm' for month and 'd' for day
.TP
.B \-s, \-\-simple
Specify a simplified output mode
.TP
.B \-u, \-\-umm_alqura
Specify to use the Umm Al-Qura calculation method (used mostly in
Saudi Arabia)
.SH BACKGROUND
The Hijri calendar is used in most of the Arab world and is the symbolic
calendar of the Islamic faithed worldwide.  This calendar is known as
the "Hijri" (based on the word "Hijrah" - denoting migration in Arabic)
to signal Prophet Mohammed's (PBUH) migration from Makkah to Medinah
on Thursday, July 15, 622 AD (Julian) or July 19, 622 AD (Gregorian).

The Islamic Hijri calendar is strictly lunar (ie. moon-based) with
twelve lunar months which do not correspond or track their solar
counterparts (the Gregorian calendar is a solar or sun-based calendar).
Lunar years and thus Hijri years are, on average, about 354 days long
resulting in a Hijri year being roughly about 11 days shorter than its
Gregorian counterpart.

There is much discussion and confusion regarding how best to track
the Hijri calendar.  A great deal of that confusion is based on the
fact that many rely on a human moon sighting to denote the start
(or end) of a month (each month of the Hijri calendar starts when
a new moon's crescent is observed or is made visible at sunset)
as opposed to using an empirical mathematic certainty.  The methods
presented in this application and its underlying ITL library are
strictly arithmetic in nature and do NOT take moon-phases into
consideration (in short, observational approximation is not used).
.SH LIMITATIONS
The Umm Al-Qura option doesn't function with pre-epoch settings.
.SH REPORTING BUGS
Report bugs on the web using http://bugs.arabeyes.org
.SH AUTHOR
Written by Nadim Shaikli as part of the Arabeyes.org project.
.SH COPYRIGHT
\fBidate\fR is subject to the GNU General Public License (GPL).
.br
Copyright \(co 2005, Arabeyes, Nadim Shaikli.
.SH SEE ALSO
\" .BR libitl (1).
.PP
The ITL library (libitl) from the Islamic Tools and Libraries project.
It is the underlying requirement for \fBidate\fR to function.  The
ITL library was created and is hosted at \fBwww.arabeyes.org\fR.


programs/itools/idate.c
/************************************************************************
 * $Id: idate.c 10994 2009-03-09 04:17:36Z thamer $
 *
 * ------------
 * Description:
 * ------------
 *  Copyright (c) 2004, Arabeyes, Nadim Shaikli
 *
 *  This is an application to primarily convert the Gregorian/Meladi
 *  date to Islamic/Hijri (from Hijri to Meladi is also offered).
 *  There are two distinct Hijri calculation conversion algorithms
 *  available (the '--umm_alqura' selecting the calculation method
 *  presented and advocated by the Saudi Umm Al-Qura University).
 *  It is envisioned that both the non-umm_alqura and the umm_alqura
 *  options will not always yield the same results.
 *
 *  NOTE: Although the non-umm_alqura algorithm used is dated, it is
 *        at this point the best implementation found.  Work will be
 *        on-going to improve its accuracy and implementation.
 *
 * -----------------
 * Revision Details:    (Updated by Revision Control System)
 * -----------------
 *  $Date: 2009-03-09 05:17:36 +0100 (Mon, 09 Mar 2009) $
 *  $Author: thamer $
 *  $Revision: 10994 $
 *  $Source$
 *
 * (www.arabeyes.org - under GPL license)
 ************************************************************************/

/* TODO:
   - Define error exit codes (standardize 'em) and associate a string
     with each to print out to let user know what went wrong
     So far:
       0 - Normal exit (no errors)
       1 - User entered invalid argument to input
       2 - User entered mututally-exclusive inputs
       3 - User entered dates beyond acceptable ranges
      10 - Can't malloc/attain memory
      20 - User invoked help
   - Add a proper 'man' page and documentation
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>             /* for strlen/strcat/etc */

/* For time_t */
#ifdef TM_IN_SYS_TIME
#include <sys/time.h>
#else
#include <time.h>
#endif

#include <itl/hijri.h>

#define PROG_NAME	"idate";


/* This holds the current date info. */
typedef struct
{
      int year;
      int month;
      int day;

} Date;


/**
  Print out the command-line usage of this application
 **/
void
usage(int leave)
{
   char *pspaces;
   char *pname = PROG_NAME;

   pspaces = (char *) malloc(strlen(pname));
   strncpy(pspaces, "                      ", strlen(pname));

   fprintf(stderr, "%s [--gregorian yyyymmdd] [--hijri yyyymmdd]\n", pname);
   fprintf(stderr, "%s                        [--simple]\n", pspaces);
   fprintf(stderr, "%s                        [--umm_alqura]\n", pspaces);
   fprintf(stderr, "%s                        [--help]\n", pspaces);

   free(pspaces);

   if (leave)
      exit(20);
}


/**
  Error printing (& possibly exit) apparatus.
 **/
void
error(int leave,
      char* err_msg)
{
   char *prog_name = PROG_NAME;

   fprintf(stderr, "[%s]: %s", prog_name, err_msg);

   if (leave)
      exit(leave);
}


/**
  Initizlize various variables/structures
 **/
void
do_init(Date *mydate,
	sDate *pdate)
{
   mydate->year		= (int) NULL;
   mydate->month	= (int) NULL;
   mydate->day		= (int) NULL;

   pdate->day		= (int) NULL;
   pdate->month		= (int) NULL;
   pdate->year		= (int) NULL;
   pdate->weekday	= (int) NULL;
   pdate->frm_numdays	= (int) NULL;
   pdate->to_numdays	= (int) NULL;
   pdate->to_numdays2	= (int) NULL;
//   pdate->units		= (char *) NULL;
   pdate->units		= "A.H";
   pdate->frm_dname	= (char *) NULL;
   pdate->frm_mname	= (char *) NULL;
   pdate->frm_dname_sh	= (char *) NULL;
   pdate->frm_mname_sh	= (char *) NULL;
   pdate->to_dname	= (char *) NULL;
   pdate->to_mname	= (char *) NULL;
   pdate->to_dname_sh	= (char *) NULL;
   pdate->to_mname_sh	= (char *) NULL;
   pdate->event		= (char **) NULL;
}


/**
  Printout the results
 **/
void
display_output(int simple_mode,
	       int using_umm_alqura,
	       Date *indate,
	       sDate *rdate)
{
   int i;

   /* See if we're only to print out a simple output */
   if (simple_mode)
   {
      printf("%2d/%2d/%4d %s", rdate->day,
	     		       rdate->month,
	     		       rdate->year,
	     		       rdate->units);
      printf(" - %s(%s) - %s(%s)\n", rdate->to_dname,
	     			     rdate->to_dname_sh,
	     			     rdate->to_mname,
	     			     rdate->to_mname_sh);
   }
   else
   {
      printf("Date Format (dd/mm/yyyy)");

      if (using_umm_alqura)
	 printf(" [using Umm-AlQura Algorithm]:\n");
      else
	 printf(":\n");

      printf("+ Input    : %2d/%2d/%4d    ", indate->day,
	  			             indate->month,
	  			             indate->year);

      printf("  - %10s(%s) - %12s(%s)\n", rdate->frm_dname,
	  			          rdate->frm_dname_sh,
	  			          rdate->frm_mname,
	  			          rdate->frm_mname_sh);

      printf("-----------------------------\n");
      printf("+ Output   : %2d/%2d/%4d %s", rdate->day,
	  			            rdate->month,
	  				    rdate->year,
      					    rdate->units);

      printf("  - %10s(%s) - %12s(%s)\n", rdate->to_dname,
	  			          rdate->to_dname_sh,
	  			          rdate->to_mname,
	  			          rdate->to_mname_sh);
   }

   /* Print out any event(s) on the specified date */
   if (rdate->event[0])
      printf("\n");

   for (i = 0; rdate->event[i] != NULL; i++)
   {
      printf("(*) Event on this Day : %s\n", rdate->event[i]);
   }
   /* Do some house cleaning - just in case */
   free(rdate->event);
}


/**
  Main procedure
 **/
int main(int argc, char *argv[])
{
   int i;
   int g_to_h		= 0;
   int h_to_g		= 0;
   int simple_mode	= 0;
   int use_umm_alqura	= 0;
   int error_lib	= 0;
   Date indate;
   sDate outdate;

   /* Current time/date specifics */
   time_t mytime;
   struct tm *t_ptr;

   do_init(&indate, &outdate);

   /* Process the command-line */
   for (i = 1; i < argc; i++)
   {
     if (strcasecmp(argv[i], "-h") == 0 ||
         strcasecmp(argv[i], "-help") == 0 ||
         strcasecmp(argv[i], "--help") == 0)
     {
        /* We really need a full-fledged help here */
        usage(1);
     }

     if (strcasecmp(argv[i], "-g") == 0 ||
         strcasecmp(argv[i], "-gregorian") == 0 ||
         strcasecmp(argv[i], "--gregorian") == 0)
     {
	g_to_h = 1;
        if (argv[i+1] != NULL)
        {
           sscanf(&(argv[i + 1][0]), "%4d", &indate.year);
           sscanf(&(argv[i + 1][4]), "%2d", &indate.month);
           sscanf(&(argv[i + 1][6]), "%2d", &indate.day);
        }
        else
        {
           error(1, "Exiting, invalid argument to --gregorian\n");
        }
     }

     if (strcasecmp(argv[i], "-hi") == 0 ||
         strcasecmp(argv[i], "-hijri") == 0 ||
         strcasecmp(argv[i], "--hijri") == 0)
     {
	h_to_g = 1;
        if (argv[i+1] != NULL)
        {
           sscanf(&(argv[i + 1][0]), "%4d", &indate.year);
           sscanf(&(argv[i + 1][4]), "%2d", &indate.month);
           sscanf(&(argv[i + 1][6]), "%2d", &indate.day);
        }
        else
        {
           error(1, "Exiting, invalid argument to --hijri\n");
        }
     }

     if (strcasecmp(argv[i], "-s") == 0 ||
         strcasecmp(argv[i], "-simple") == 0 ||
         strcasecmp(argv[i], "--simple") == 0)
     {
	 simple_mode = 1;
     }

     if (strcasecmp(argv[i], "-u") == 0 ||
         strcasecmp(argv[i], "-umm_alqura") == 0 ||
         strcasecmp(argv[i], "--umm_alqura") == 0)
     {
	 use_umm_alqura = 1;
     }
   }

   /* Make sure user knows what they are doing */
   if (g_to_h && h_to_g)
      error(2, "Exiting, can't define both Gregorian and Hijri\n");

   /* Go with default current date if not user specifications */
   if (!g_to_h && !h_to_g)
   {
      g_to_h = 1;

      /* Get current time structure */
      time(&mytime);

      t_ptr = localtime(&mytime);

      /* Set current time values */
      indate.day   = t_ptr->tm_mday;
      indate.month = t_ptr->tm_mon  + 1;
      indate.year  = t_ptr->tm_year + 1900;
   }

   /* Inspect user's input or default settings - within range ? */
   if ((indate.month > 12) || (indate.month < 1))
      error(3, "Exiting, input 'month' not within allowable range\n");

   if ((indate.day > 31) || (indate.day < 1))
      error(3, "Exiting, input 'day' not within allowable range\n");

   /* Do the actual conversions */
   if (g_to_h)
   {
      /* Dealing with Gregorian to Hijri conversion */
      if (use_umm_alqura)
	 G2H(&outdate, indate.day, indate.month, indate.year);
      else
	 error_lib = h_date(&outdate, indate.day, indate.month, indate.year);
   }
   else
   {
      /* Dealing with Hijri to Gregorian conversion */
      if (use_umm_alqura)
	 H2G(&outdate, indate.day, indate.month, indate.year);
      else
	 error_lib = g_date(&outdate, indate.day, indate.month, indate.year);
   }

   /* Let user know if the library freaked-out and returned with an error */
   if (error_lib)
   {
      char msg[80];

      sprintf(msg, "Exiting, error code %d generated by library call\n",
	      error_lib);
      error(error_lib, msg);
   }

   /* Spill-out the results */
   display_output(simple_mode, use_umm_alqura, &indate, &outdate);

   return(0);
}
